polkadot/runtime/parachains/src/paras/mod.rs

```diff
@@ -618,6 +669,13 @@ pub mod pallet {
 		/// Runtime hook for when a parachain head is updated.
 		type OnNewHead: OnNewHead;
 
+		/// A type that performs custom logic to determine whether a code upgrade is allowed to be
+		/// performed.
+		type PreCodeUpgrade: PreCodeUpgrade;
+
+		/// Type that executes some custom logic upon a successful code upgrade.
+		type OnCodeUpgrade: OnCodeUpgrade;
```

Comment:
Maybe we should just merge both and also the traits.

The trait could maybe just be `CodeUpgradeHandler`. 

(the comment is not a strict requirement)